Removed PreOutInterface and PostOutInterface states

anothermattbrown opened the pull request:
These have been replaced by PreOutEdgePostNat and PostOutEdge. The key difference is that the new states are specific to the destination node/interface pair of the edge. This means one less join is necessary -- a particularly problematic join in the presence of NAT.

arifogel commented:
High level: I complained about some stuff that is probably out of scope, and some stuff that is super worrisome. Let's chat.

---

Reviewed 21 of 21 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[projects/batfish/src/main/java/org/batfish/z3/ReachabilityQuerySynthesizer.java, line 177 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6ivXjz5Wa3m2vagl1l:-L6ivXk-9IuXvhS_2u0K:bxcej3t) ([raw file](https://github.com/batfish/batfish/blob/eb4c7f0251bb706dd4b4332688ff55bdb43b8b27/projects/batfish/src/main/java/org/batfish/z3/ReachabilityQuerySynthesizer.java#L177)):*
> ```Java
>     //        .map(NodeTransit::new)
>     //        .map(NotExpr::new)
>     //        .forEach(queryPreconditionPreTransformationStates::add);
> ```

Is there a ticket to fix this?

---

*[projects/batfish/src/main/java/org/batfish/z3/SynthesizerInput.java, line 74 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6iwH7R8dm3owahVCFJ:-L6iwH7SAjPyhwrPR360:b73iypl) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/SynthesizerInput.java#L74)):*
> ```Java
>    * Mapping: hostname -> interface -> [(preconditionPreTransformationState, transformationToApply)]
>    */
>   Map<String, Map<String, List<Entry<AclPermit, BooleanExpr>>>> getSourceNats();
> ```

Why not `Optional<AclPermit>`? Are you using some always-permit ACL when there is none in the `SourceNat`?

---

*[projects/batfish/src/main/java/org/batfish/z3/SynthesizerInputImpl.java, line 617 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6ix_ZdD7Yjj3ULUm5n:-L6ix_ZdD7Yjj3ULUm5o:b-ixedp) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/SynthesizerInputImpl.java#L617)):*
> ```Java
>                                             IpAccessList acl = sourceNat.getAcl();
>                                             AclPermit preconditionPreTransformationState =
>                                                 new AclPermit(hostname, acl.getName());
> ```

How does this not cause NPE when `sourceNat` has no acl?

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 210 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6izNj18VcH7yi2IMC9:-L6izNj18VcH7yi2IMCA:b9uuz20) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L210)):*
> ```Java
>                                           : ImmutableSet.of();
>                                   return new BasicRuleStatement(
>                                       new NotExpr(lineCriteria),
> ```

Wow how embarrassing :P

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 449 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j06zA9iMZTnFCNgcq:-L6j06zA9iMZTnFCNgcr:b5upqia) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L449)):*
> ```Java
>                     ImmutableSet.of(new AclDeny(node1, outAcl));
>                 _rules.add(
>                     new TransformedBasicRuleStatement(
> ```

This looks wrong. There is no transformation happening here. Under the assumption we have a packet that made it past source NAT (and has been transformed already), all we have to do is say that that (already transformed) packet gets dropped.
Thus, we should use a regular `BasicRuleStatement`, with two precondition states: PreOutEdgePostNat` and `AclDeny`.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 701 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j0XFhEdWWLM-X1Ch4:-L6j0XFhEdWWLM-X1Ch5:bb91tdu) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L701)):*
> ```Java
>                       ? ImmutableSet.of()
>                       : ImmutableSet.of(new AclPermit(node1, outAcl));
>               return new TransformedBasicRuleStatement(
> ```

There should not be any transformation here either. This should be a `BasicRuleStatement` like the one I suggested for `NodeDropAclOut`.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 761 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j19fECw_BMup0haZy:-L6j19fECw_BMup0haZz:b-u7qy2w) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L761)):*
> ```Java
>         .stream()
>         .map(Entry::getKey)
>         .map(aclPermit -> new AclDeny(aclPermit.getHostname(), aclPermit.getAcl()))
> ```

Not every source NAT has an ACL. I don't yet see how that case was handled. See comment about removing `Optional`.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 765 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j1aXL9A34WkiIBpNB:-L6j1aXL9A34WkiIBpNC:bmshwwh) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L765)):*
> ```Java
> 
>     _rules.add(
>         new TransformationRuleStatement(
> ```

You are not yet removing `TransformationRuleStatement`?

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 905 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j2PFNCUgj6Gl44_mH:-L6j2PFNCUgj6Gl44_mI:bqzwsb2) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L905)):*
> ```Java
>                                 fibConditionsByOutInterfaceEntry -> {
>                                   // TODO what's going on here?
>                                   // can we use flow sinks for this?
> ```

Rules for routing out an interface are not dependent on whether that interface is a flow sink. We need to guarantee these rules are generated for both flow sink interfaces and interfaces on edges.

---

*[projects/batfish/src/test/java/org/batfish/z3/expr/visitors/BoolExprTransformerTest.java, line 81 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j3N7l6fPnRZcDNOJY:-L6j3N7l6fPnRZcDNOJZ:b-7f1sqx) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/test/java/org/batfish/z3/expr/visitors/BoolExprTransformerTest.java#L81)):*
> ```Java
>   public void setup() {
>     _basicStateExpr = Accept.INSTANCE;
>     _transformationStateExpr = new PreOutEdgePostNat("host1", "interface1", "host2", "interface2");
> ```

Should this be a `TransformationStateExpr`? If you are trying to defer removing `TransformationStateExpr`, I guess so, but it makes everything dirty. I suppose this is related to my objections in `DefaultTransitionGenerator`. I am leaving them for reference.

---

*[projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java, line 825 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j4OHU76LSXRszW78T:-L6j4OHU76LSXRszW78U:bmk2puz) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java#L825)):*
> ```Java
> 
>     // PostOutFlowSinkInterface
>     for (String node1 : new String[] {NODE1, NODE2}) {
> ```

Variable name is weird here.

---

*[projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java, line 826 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j4QXk08MSzrHV3PwI:-L6j4QXk08MSzrHV3PwJ:bmk2puz) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java#L826)):*
> ```Java
>     // PostOutFlowSinkInterface
>     for (String node1 : new String[] {NODE1, NODE2}) {
>       for (String node1Interface : new String[] {INTERFACE1, INTERFACE2}) {
> ```

Variable name is weird here.

---

*[projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java, line 966 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j4e2BE59PCTX18Yuk:-L6j4e2BE59PCTX18Yul:b970je1) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java#L966)):*
> ```Java
>                         ImmutableList.of(
>                             Maps.immutableEntry(
>                                 // random guess
> ```

This is exactly what I was complaining about regarding removal of `Optional`. You had to throw away a test that now fails.

---

*[projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java, line 1005 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j58452UCHiVGl2MtB:-L6j58452UCHiVGl2MtC:beaisju) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java#L1005)):*
> ```Java
>     assertThat(
>         rules,
>         not(
> ```

Let's have a high-bandwidth conversation about what is going on now.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1019#-:-L6j5PvhCgC_TRKOxIBb:b-5q6951)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/ReachabilityQuerySynthesizer.java, line 177 at r1](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6ivXjz5Wa3m2vagl1l:-L6r9CpvK0kqpUWdNasZ:b-rytyoo) ([raw file](https://github.com/batfish/batfish/blob/eb4c7f0251bb706dd4b4332688ff55bdb43b8b27/projects/batfish/src/main/java/org/batfish/z3/ReachabilityQuerySynthesizer.java#L177)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Is there a ticket to fix this?
</blockquote></details>

yes, #993.

---

*[projects/batfish/src/main/java/org/batfish/z3/SynthesizerInput.java, line 74 at r1](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6iwH7R8dm3owahVCFJ:-L6r9JCiHlyl3zelZtVr:b35dpp) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/SynthesizerInput.java#L74)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Why not `Optional<AclPermit>`? Are you using some always-permit ACL when there is none in the `SourceNat`?
</blockquote></details>

That was the plan, with the goal of improving clarity and simplicity. I see that I introduced a bug though -- thanks for the catch.

We should make default behaviors explicit. This makes the behavior clear and makes the code simpler by reducing the number of cases we have to consider. Ideally this should be done in one place, and as early as possible. 

---

*[projects/batfish/src/main/java/org/batfish/z3/SynthesizerInputImpl.java, line 617 at r1](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6ix_ZdD7Yjj3ULUm5n:-L6rBDlwBtQCrNlYfLLI:b-cgmwhj) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/SynthesizerInputImpl.java#L617)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

How does this not cause NPE when `sourceNat` has no acl?
</blockquote></details>

That should be disallowed. Will fix. 

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 449 at r1](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6j06zA9iMZTnFCNgcq:-L6rBqH0Vwl4CPe-Mt5X:b-yruphd) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L449)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

This looks wrong. There is no transformation happening here. Under the assumption we have a packet that made it past source NAT (and has been transformed already), all we have to do is say that that (already transformed) packet gets dropped.
Thus, we should use a regular `BasicRuleStatement`, with two precondition states: PreOutEdgePostNat` and `AclDeny`.
</blockquote></details>

I agree, but we can't do that yet because `PreOutEdgePostNat` is a `TransformationStateExpr`.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 701 at r1](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6j0XFhEdWWLM-X1Ch4:-L6rC8z3YikdJR86K5uo:b-d22waz) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L701)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

There should not be any transformation here either. This should be a `BasicRuleStatement` like the one I suggested for `NodeDropAclOut`.
</blockquote></details>

Same response as above.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 761 at r1](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6j19fECw_BMup0haZy:-L6rCIONb3s0mqvy2egW:bjfjr1a) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L761)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Not every source NAT has an ACL. I don't yet see how that case was handled. See comment about removing `Optional`.
</blockquote></details>

I will make every source NAT have an ACL.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 765 at r1](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6j1aXL9A34WkiIBpNB:-L6rCO8-8OPAn-SjiJAq:bs5a1q9) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L765)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

You are not yet removing `TransformationRuleStatement`?
</blockquote></details>

Nope not yet. Gettings the new states and transitions in place first.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 905 at r1](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6j2PFNCUgj6Gl44_mH:-L6rCXMdwUexPcv5B5Oh:b-6c994k) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L905)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Rules for routing out an interface are not dependent on whether that interface is a flow sink. We need to guarantee these rules are generated for both flow sink interfaces and interfaces on edges.
</blockquote></details>

Ok, let's discuss this

---

*[projects/batfish/src/test/java/org/batfish/z3/expr/visitors/BoolExprTransformerTest.java, line 81 at r1](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6j3N7l6fPnRZcDNOJY:-L6rCl60TWgxPH60Cx71:b-jiz9gq) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/test/java/org/batfish/z3/expr/visitors/BoolExprTransformerTest.java#L81)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Should this be a `TransformationStateExpr`? If you are trying to defer removing `TransformationStateExpr`, I guess so, but it makes everything dirty. I suppose this is related to my objections in `DefaultTransitionGenerator`. I am leaving them for reference.
</blockquote></details>

Yes, I'm tackling one thing at a time. Removing TransformationStateExpr is next.

---

*[projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java, line 966 at r1](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6j4e2BE59PCTX18Yuk:-L6rD0cn8jmFYtTFmxmX:b-numin0) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java#L966)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

This is exactly what I was complaining about regarding removal of `Optional`. You had to throw away a test that now fails.
</blockquote></details>

We shouldn't use Optional.empty or null -- just give the default value. My intent was to remove a case that needs testing, not a test for a valid case.

---

*[projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java, line 1005 at r1](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6j58452UCHiVGl2MtB:-L6rF2ZGA2Qg_tOFj4Rw:b-tdqrg6) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java#L1005)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Let's have a high-bandwidth conversation about what is going on now.
</blockquote></details>

This is an inadequate test, but the idea is that PreOutEdgePostNat should not imply NodeDropAclOut without AclDeny.  I added it because my bug Friday was that this rule was being generated. In general, it is not sufficient to test only that what should be generated is -- we should also test that what shouldn't be isn't. In this case it's probably best to use a full-set matcher rather than a bunch of hasItems.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 449 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j06zA9iMZTnFCNgcq:-L6rfFaP-8MlVOXdpYJO:b-2ic57d) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L449)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

I agree, but we can't do that yet because `PreOutEdgePostNat` is a `TransformationStateExpr`.
</blockquote></details>

OK. You are right - this is out of scope for now.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 701 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j0XFhEdWWLM-X1Ch4:-L6rfL41BcP8Iq4X6huW:b-2ic57d) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L701)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Same response as above.
</blockquote></details>

OK. You are right - this is out of scope for now.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 761 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j19fECw_BMup0haZy:-L6rfMvRC_xBlE-XufrU:b2nhktp) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L761)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

I will make every source NAT have an ACL.
</blockquote></details>

OK.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 765 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j1aXL9A34WkiIBpNB:-L6rfRQt7dzgwLXBlT2j:b-2ic57d) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L765)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Nope not yet. Gettings the new states and transitions in place first.
</blockquote></details>

OK. You are right - this is out of scope for now.

---

*[projects/batfish/src/test/java/org/batfish/z3/expr/visitors/BoolExprTransformerTest.java, line 81 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j3N7l6fPnRZcDNOJY:-L6rfZPyB_KcZm9xVcOJ:b-2ic57d) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/test/java/org/batfish/z3/expr/visitors/BoolExprTransformerTest.java#L81)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Yes, I'm tackling one thing at a time. Removing TransformationStateExpr is next.
</blockquote></details>

OK. You are right - this is out of scope for now.

---

*[projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java, line 1005 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j58452UCHiVGl2MtB:-L6rfvQb9DsxdhEkv9Z7:bo5u7g1) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java#L1005)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

This is an inadequate test, but the idea is that PreOutEdgePostNat should not imply NodeDropAclOut without AclDeny.  I added it because my bug Friday was that this rule was being generated. In general, it is not sufficient to test only that what should be generated is -- we should also test that what shouldn't be isn't. In this case it's probably best to use a full-set matcher rather than a bunch of hasItems.
</blockquote></details>

You are right about the need for negative tests. To avoid the full universe of negative tests, one solution is to simply compare to exact expected contents. Unfortunately, when such a test fails, it is very difficult to debug. Let's talk offline about best way to approach this.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/SynthesizerInput.java, line 74 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6iwH7R8dm3owahVCFJ:-L6s34y84zAWK6C04_4_:blrsazk) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/SynthesizerInput.java#L74)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

That was the plan, with the goal of improving clarity and simplicity. I see that I introduced a bug though -- thanks for the catch.

We should make default behaviors explicit. This makes the behavior clear and makes the code simpler by reducing the number of cases we have to consider. Ideally this should be done in one place, and as early as possible. 
</blockquote></details>

Fine once your extra enabled ACL is merged.

---

*[projects/batfish/src/main/java/org/batfish/z3/SynthesizerInputImpl.java, line 617 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6ix_ZdD7Yjj3ULUm5n:-L6s3E1U6tN0amEDwnYx:blrsazk) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/SynthesizerInputImpl.java#L617)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

That should be disallowed. Will fix. 
</blockquote></details>

Fine once your extra enabled ACL is merged.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 905 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j2PFNCUgj6Gl44_mH:-L6s3HhsC2hcTEpM15Xy:bi52cjd) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L905)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Ok, let's discuss this
</blockquote></details>

We didn't discuss this. What do you want to do with it?

---

*[projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java, line 966 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j4e2BE59PCTX18Yuk:-L6s3NcdB2K0awidBkmf:b-uyo4rn) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java#L966)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

We shouldn't use Optional.empty or null -- just give the default value. My intent was to remove a case that needs testing, not a test for a valid case.
</blockquote></details>

Awaiting update after extra enabled ACL is merged.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Reviewed 5 of 5 files at r2.
Review status: 21 of 22 files reviewed at latest revision, 6 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 787 at r2](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6s5DPM0Ezla3az2O5x:-L6s5DPM0Ezla3az2O5y:b-uz8vyf) ([raw file](https://github.com/batfish/batfish/blob/f0ba4c98a5e3489da3a1b67b6c2da2b8444dfb92/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L787)):*
> ```Java
>         _input
>             .getSourceNats()
>             .getOrDefault(node1, ImmutableMap.of())
> ```

Is this still necessary?

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 788 at r2](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6s5S039ZGnXpuuT7NY:-L6s5S039ZGnXpuuT7NZ:b-uz8vyf) ([raw file](https://github.com/batfish/batfish/blob/f0ba4c98a5e3489da3a1b67b6c2da2b8444dfb92/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L788)):*
> ```Java
>             .getSourceNats()
>             .getOrDefault(node1, ImmutableMap.of())
>             .getOrDefault(iface1, ImmutableList.of());
> ```

Is this still necessary?

---

*[projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java, line 1767 at r2](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6s66SE5LvqX75TlCKT:-L6s66SFC9sgSXWgB0CC:b4n0tio) ([raw file](https://github.com/batfish/batfish/blob/f0ba4c98a5e3489da3a1b67b6c2da2b8444dfb92/projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java#L1767)):*
> ```Java
>             new PreOutEdgePostNat(NODE1, INTERFACE1, NODE2, INTERFACE2));
> 
>     assertThat(rules, contains(permitRule, denyRule));
> ```

`contains` -> `containsInAnyOrder`?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Review status: 21 of 22 files reviewed at latest revision, 9 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 905 at r1](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6j2PFNCUgj6Gl44_mH:-L6s7JMufVZKEwRBjmXm:bi8ec90) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L905)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

We didn't discuss this. What do you want to do with it?
</blockquote></details>

Maybe just factor out that predicate into a method -- what would be a good name? Something positive, like interfaceIsRealForSomeNotionOfRealness or interfaceHasSomeDesirableProperty would be good

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 787 at r2](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6s5DPM0Ezla3az2O5x:-L6s6i_4VOoQLCHx2bvi:bsc24yw) ([raw file](https://github.com/batfish/batfish/blob/f0ba4c98a5e3489da3a1b67b6c2da2b8444dfb92/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L787)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Is this still necessary?
</blockquote></details>

I think so -- some nodes may not have any source NATs

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 788 at r2](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6s5S039ZGnXpuuT7NY:-L6s73CE1cP34pjoZhVd:b-izzoat) ([raw file](https://github.com/batfish/batfish/blob/f0ba4c98a5e3489da3a1b67b6c2da2b8444dfb92/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L788)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Is this still necessary?
</blockquote></details>

I think so -- some interfaces may not have any source NATs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Review status: 21 of 22 files reviewed at latest revision, 9 unresolved discussions.

---

*[projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java, line 825 at r1](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6j4OHU76LSXRszW78T:-L6sEObnJmhrj_Cz8Qv_:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java#L825)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Variable name is weird here.
</blockquote></details>

Done.

---

*[projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java, line 826 at r1](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6j4QXk08MSzrHV3PwI:-L6sEOyWWktl2KSS9-TO:b-pyqwih) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java#L826)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Variable name is weird here.
</blockquote></details>

Done.....

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Review status: 21 of 22 files reviewed at latest revision, 9 unresolved discussions.

---

*[projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java, line 1767 at r2](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6s66SE5LvqX75TlCKT:-L6sERhiB4Xl4-q4SZwS:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/f0ba4c98a5e3489da3a1b67b6c2da2b8444dfb92/projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java#L1767)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

`contains` -> `containsInAnyOrder`?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 905 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j2PFNCUgj6Gl44_mH:-L6sKFbl26tJ77b8r_w3:bcj0nvh) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L905)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Maybe just factor out that predicate into a method -- what would be a good name? Something positive, like interfaceIsRealForSomeNotionOfRealness or interfaceHasSomeDesirableProperty would be good
</blockquote></details>

Any interface that can be used as a next hop interface for forwarding needs to be processed here.
This excludes:
- interfaces without any IP address
- interfaces that are off (should already be filtered by this point)
- loopback interfaces

It DOES include:
- flow sinks (interfaces not attached to an edge that are not excluded above)
- null interface

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 905 at r1](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6j2PFNCUgj6Gl44_mH:-L6sLN9CUKhs-5zU5nnz:berx25q) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L905)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Any interface that can be used as a next hop interface for forwarding needs to be processed here.
This excludes:
- interfaces without any IP address
- interfaces that are off (should already be filtered by this point)
- loopback interfaces

It DOES include:
- flow sinks (interfaces not attached to an edge that are not excluded above)
- null interface

</blockquote></details>

Ok, so maybe `isValidNextHopInterface` would be a good name?
It looks like null interfaces are excluded. Is that a bug?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 905 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j2PFNCUgj6Gl44_mH:-L6sMbmN-KVBTyBcmTCP:b-m0tv5p) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L905)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Ok, so maybe `isValidNextHopInterface` would be a good name?
It looks like null interfaces are excluded. Is that a bug?
</blockquote></details>

It is a bug if there is no other way to get to NodeDropNullRoute

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Review status: 21 of 23 files reviewed at latest revision, 4 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 905 at r1](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6j2PFNCUgj6Gl44_mH:-L6sOzztBdDxnXllr3RF:b-7w5gue) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L905)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

It is a bug if there is no other way to get to NodeDropNullRoute
</blockquote></details>

Nope I see it -- visitNodeDropNullRoute

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 900 at r5](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6s_r-g9Zri2hNjfTqj:-L6s_r-h41zeR_kfxDXS:bdnv9ha) ([raw file](https://github.com/batfish/batfish/blob/21a19bc6e930e58434b7f8bff20a8eef77cba257/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L900)):*
> ```Java
>                                 fibConditionsByOutInterfaceEntry -> {
>                                   String outInterface = fibConditionsByOutInterfaceEntry.getKey();
>                                   // Loopback and Null Interfaces are handled in
> ```

Please use multiline comment style:
```
/*
 * comment line 1
 * comment line 2
 */
```

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 900 at r5](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6s_r-g9Zri2hNjfTqj:-L6sbbhbujKWZ0UFAfpN:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/21a19bc6e930e58434b7f8bff20a8eef77cba257/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L900)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Please use multiline comment style:
```
/*
 * comment line 1
 * comment line 2
 */
```
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 900 at r5](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6s_r-g9Zri2hNjfTqj:-L6sbqkr8aDj_yigB83K:brp34ok) ([raw file](https://github.com/batfish/batfish/blob/21a19bc6e930e58434b7f8bff20a8eef77cba257/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L900)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

So close! Please no text after `/*` until next line.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Review status: 22 of 23 files reviewed at latest revision, 4 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 900 at r5](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6s_r-g9Zri2hNjfTqj:-L6scrE_WTtl4gQxoPTn:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/21a19bc6e930e58434b7f8bff20a8eef77cba257/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L900)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

So close! Please no text after `/*` until next line.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/SynthesizerInputImpl.java, line 617 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6ix_ZdD7Yjj3ULUm5n:-L6sdTsp8LQuJdZxNvTu:bntumkk) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/SynthesizerInputImpl.java#L617)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Fine once your extra enabled ACL is merged.
</blockquote></details>

(Inserting extra comment cause Reviewable in hopes of clearing unreplied comment)

---

*[projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java, line 966 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j4e2BE59PCTX18Yuk:-L6sdK4SE1tDEXhjEfCH:bntumkk) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java#L966)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Awaiting update after extra enabled ACL is merged.
</blockquote></details>

(Inserting extra comment cause Reviewable in hopes of clearing unreplied comment)

---

*[projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java, line 1005 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j58452UCHiVGl2MtB:-L6sdTHW79J9vtqiYC4H:bntumkk) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java#L1005)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

You are right about the need for negative tests. To avoid the full universe of negative tests, one solution is to simply compare to exact expected contents. Unfortunately, when such a test fails, it is very difficult to debug. Let's talk offline about best way to approach this.
</blockquote></details>

(Inserting extra comment cause Reviewable in hopes of clearing unreplied comment)

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/SynthesizerInputImpl.java, line 617 at r1](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6ix_ZdD7Yjj3ULUm5n:-L6se1jIP5ifG3PrqLwp:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/main/java/org/batfish/z3/SynthesizerInputImpl.java#L617)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

(Inserting extra comment cause Reviewable in hopes of clearing unreplied comment)
</blockquote></details>

Done.

---

*[projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java, line 966 at r1](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6j4e2BE59PCTX18Yuk:-L6se3lgN8XzcsgXF25t:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java#L966)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

(Inserting extra comment cause Reviewable in hopes of clearing unreplied comment)
</blockquote></details>

Done.

---

*[projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java, line 1005 at r1](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6j58452UCHiVGl2MtB:-L6se5NpOVLY_NbY20Qw:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java#L1005)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

(Inserting extra comment cause Reviewable in hopes of clearing unreplied comment)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java, line 1005 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1019#-L6j58452UCHiVGl2MtB:-L6shN2E8ItoQxou1kFk:b-67dp1u) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java#L1005)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Awaiting actual changes.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java, line 1005 at r1](https://reviewable.io:443/reviews/batfish/batfish/1019#-L6j58452UCHiVGl2MtB:-L6snmJmhxjiysK_1m_N:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/1ec946f4f7e545f81bf3c66824d172c64fd4a051/projects/batfish/src/test/java/org/batfish/z3/state/visitors/DefaultTransitionGeneratorTest.java#L1005)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Awaiting actual changes.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Run `.travis/fix_java_format.sh`

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1019#-:-L6sp1Gj0mNQ5_4EuirN:beadol9)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1019)*
<!-- Sent from Reviewable.io -->
